### ISSUE TITLE
Time comparison between PyTorch and TensorFlow

### ISSUE THREAD

**Issue by irwinherrmann** [kmaninis/OSVOS-PyTorch#16]
So I did some time benchmarks and found that about 10 epochs takes the following amount of time:

Env:
OS: Ubuntu 16.04
GPU: Dual 1080 TIs with HB SLI bridge

TensorFlow 1.9: 0.85 seconds
PyTorch 0.5:      2.32 seconds

Is this consistent with what you've observed? I'm trying to figure out which of the following is true: 1) this is a side effect of my installation; 2) the TensorFlow version is much more optimized; 3) there is an underlying side effect that makes TensorFlow faster than PyTorch for this network.

**Comment by kmaninis** [comment]
Hi,

Just to make sure you are comparing apples to apples.
In PyTorch, the default setting runs 5 forward passes before doing a backward pass (nAveGrad parameter). This is not the case for the demo in the TensorFlow repo. As a matter of fact, the results of the paper were obtained by setting number of iterations to 2000 and nAveGrad to 5 (the TF demo has 500 and 1, respectively).  Apart from that, PyTorch runtime should be at least equal to the TF one.

Please let us know once you check.

**Comment by scaelles** [comment]
Hello,
As my colleague suggests, I would recommend you setting `iter_mean_grad` [here](https://github.com/scaelles/OSVOS-TensorFlow/blob/6ef61d1f523296b95974a6dafac01e94bf2fde7d/osvos_demo.py#L57) to 5 and then compare both frameworks for a certain number of iterations.

Let us know which is the result.

**Comment by irwinherrmann** [comment]
So I tried a couple of different things. First I tried Sergi's suggestion, with iter_mean_grad as 5 in the location. The average time per epoch (over 500 epochs):

PyTorch:      0.222
TensorFlow: 0.376

I also tried nAveGrad [here](https://github.com/kmaninis/OSVOS-PyTorch/blob/master/train_online.py#L40) as 1. The average time per epoch (over 500 epochs):

PyTorch:      0.225
TensorFlow: 0.0896

I did a breakdown on the PyTorch steps and apparently the forward passes don't take much time - it's the calculation of cross entropy that dominates the epoch time. There seems to be a very strange time trade-off happening here where forward passes take more time in TensorFlow than PyTorch.

In your experience, was there a large performance difference in terms of final loss between iter_mean_grad = 5 vs iter_mean_grad = 1? The fine_tune train loss seems to be driven down much faster with iter_mean_grad = 1, but ofc that could be overfitting.

**Comment by kmaninis** [comment]
Hi, thanks for sharing the stats.

To answer your question, it is normal that the loss goes down much faster with iter_mean_grad=1, since there are 5 times more updates. There is a small performance difference, but less than 1%, in our experiments.

There are more tricks to keep the running time down, that we haven't used, like increasing the batch size, and also cudnn.benchmark=True in PyTorch.

**Comment by irwinherrmann** [comment]
Is the 1% performance difference after the 2000 iterations for both? Or does that percentage difference also hold for 500 iterations? This is 1% on the test set, right?

I guess the thing that I find strange is the large difference in runtime for the cross entropy calculation between PyTorch and TensorFlow. I guess I'll spend some time looking into it.

**Comment by scaelles** [comment]
Maybe the way we implemented it in Pytorch is not optimum. Let us know if you find something and feel free to make a pull request!
